### ISSUE TITLE
tap0 interface is created with ipv6 only?

### ISSUE THREAD

**Issue by mattgray** [mirage/mirage#363]
I'm trying to run the example services in mirage-tcpip, but I'm unable to make them respond on ipv4
Running Ubuntu 14.04.1 LTS

steps:

```
$ git clone git@github.com:mirage/mirage-tcpip.git
$ cd mirage-tcpip/examples
$ env ADDR=static mirage configure
$ make depend
$ make
$ sudo ./mir-services

plugging into tap0 with mac c2:9d:56:19:d7:2c..
Netif: connect tap0
Manager: connect
Manager: configuring
Manager: Interface to 10.0.0.2 nm 255.255.255.0 gw [10.0.0.1]

ARP: sending gratuitous from 10.0.0.2
Manager: configuration done
```

(with mir-services running)

```
$ ifconfig tap0
tap0      Link encap:Ethernet  HWaddr ee:96:a4:78:ae:ce  
          inet6 addr: fe80::ec96:a4ff:fe78:aece/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:1 errors:0 dropped:0 overruns:0 frame:0
          TX packets:8 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:500 
          RX bytes:42 (42.0 B)  TX bytes:648 (648.0 B)
```

I'm not able to add the ipv4 route to 10.0.0.2

Matt

**Comment by mattgray** [comment]
Here's the generated main.ml
```
examples|master⚡ ⇒ cat main.ml
(* Generated by Mirage (Tue, 10 Feb 2015 14:15:54 GMT). *)

open Lwt

let _ = Printexc.record_backtrace true

module Console = Console_unix

let console1 () =
  Console.connect "0"

module M2 = Services.Main(Console)

let t2 = console1

let clock () = return (`Ok ())

let net_tap0 () =
  Netif.connect "tap0"

let random () = return (`Ok ())

module Stackv41 = struct
  module E = Ethif.Make(Netif)
  module I = Ipv4.Make(E)
  module U = Udp.Make(I)
  module T = Tcp.Flow.Make(I)(OS.Time)(Clock)(Random)
  module S = Tcpip_stack_direct.Make(Console)(OS.Time)(Random)(Netif)(E)(I)(U)(T)
  include S
end

let stackv41 () =
  console1 () >>= function
  | `Error _    -> fail (Failure "console1")
  | `Ok console ->
  net_tap0 () >>= function
  | `Error e      ->
    fail (Failure (Mirage_runtime.string_of_network_init_error "net_tap0" e))
  | `Ok interface ->
  let config = {
    V1_LWT.name = "stackv41";
    console; interface;
    mode = `IPv4 (Ipaddr.V4.of_string_exn "10.0.0.2", Ipaddr.V4.of_string_exn "255.255.255.0", [Ipaddr.V4.of_string_exn "10.0.0.1"]);
  } in
  Stackv41.connect config

module M1 = Services.Main(Console)(Stackv41)

let t1 () =
  console1 () >>= function
  | `Error e -> fail (Failure "console1")
  | `Ok console1 ->
  stackv41 () >>= function
  | `Error e -> fail (Failure "stackv41")
  | `Ok stackv41 ->
  M1.start console1 stackv41

let () =
  OS.Main.run (join [t1 ()])
```

**Comment by mattgray** [comment]
```
$ mirage --version
2.0.0

$ opam list | grep mirage
mirage                  2.2.0  The Mirage library operating system
mirage-clock-unix       1.0.0  A Mirage-compatible Clock library for Unix
mirage-clock-xen        1.0.0  A Mirage-compatible Clock library for Xen
mirage-conduit          2.0.0  Virtual package for the Mirage Conduit transports
mirage-console          2.1.0  A Mirage-compatible Console library for Xen and Unix
mirage-dns              2.0.0  Virtual package for the Mirage DNS transports
mirage-fs-unix          1.1.3  Mirage filesystem passthrough driver for Unix
mirage-http             2.1.0  Mirage HTTP client and server driver
mirage-net-unix         1.1.1  Ethernet network driver for Mirage, using tuntap
mirage-profile            0.4  Collect profiling information
mirage-types            2.2.0  Module type definitions for Mirage-compatible applications
mirage-types-lwt        2.2.0  Lwt module type definitions for Mirage-compatible applications
mirage-unix             2.1.2  Mirage OS library for Unix compilation
mirage-xen              2.1.2  Mirage OS library for Xen compilation
mirage-xen-minios       0.6.0  Xen MiniOS guest operating system library
```

**Comment by mattgray** [comment]
I can make it work, if I manually configure tap0 before running the example:
```
$ sudo tunctl        
Set 'tap0' persistent and owned by uid 0
$sudo ifconfig tap0 inet 10.0.0.1 netmask 255.255.255.0
$ ifconfig tap0
tap0      Link encap:Ethernet  HWaddr 1a:b4:80:a0:ae:2e  
          inet addr:10.0.0.1  Bcast:10.0.0.255  Mask:255.255.255.0
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:500 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
```

When I run the mir-service now, I get 
```
$ sudo ./mir-services
plugging into tap0 with mac c2:9d:56:19:d7:2c..
Netif: connect tap0
Manager: connect
Manager: configuring
Manager: Interface to 10.0.0.2 nm 255.255.255.0 gw [10.0.0.1]

ARP: sending gratuitous from 10.0.0.2
Manager: configuration done
ARP responding to: who-has 10.0.0.2?
ARP: transmitting probe -> 10.0.0.1
ARP: updating 10.0.0.1 -> 1a:b4:80:a0:ae:2e
```
and ping / other examples work

**Comment by samoht** [comment]
Someone with more knowledge than me on the gory network details (@magnuss?) could maybe help?

**Comment by hannesm** [comment]
TL;DR: IMHO this is a non-issue

The MirageOS unikernel requests a tap interface from the host system, and configures its network configuration on there.  The other end of the tap interface, in the host system, is not under control of the MirageOS unikernel, and thus it must be configured by someone else (an automated script or manually).  It might be worth to add a log message during startup in such a configuration (target = unix, net = direct, dhcp = no)!?

**Comment by mattgray** [comment]
> TL;DR: IMHO this is a non-issue

I agree
